File: matlab/demo_barbershop.m
%% demo_barbershop.m - Demo of Multi-F0 Tracker with Synthetic Barbershop Quartet
%
% Generates a synthetic barbershop quartet (A dominant 7th in just intonation)
% with staggered voice entries/exits
% and then runs the Multi-F0 tracker
%
% Chord voicing:
%   Bass:     A2  = 110.0 Hz  (1:1 root)
%   Baritone: C#3 = 137.5 Hz  (5:4 major third)
%   Lead:     E3  = 165.0 Hz  (3:2 perfect fifth)
%   Tenor:    G3  = 192.5 Hz  (7:4 harmonic seventh)

clear; close all; clc;

%% Configuration
sr = 22050;
duration = 9.5;  % seconds

% Voice parameters: [frequency, entry_time, exit_time]
voices = dictionary( ...
    "Bass",     [110.0,  0.5, 9.0], ...
    "Baritone", [137.5,  1.5, 8.0], ...
    "Lead",     [165.0,  2.5, 7.0], ...
    "Tenor",    [192.5,  3.5, 6.0]);

%% Generate synthetic audio
fprintf("🎵 Generating synthetic barbershop quartet...\n\n");

t = (0:round(duration*sr)-1)' / sr;
audio = zeros(size(t));

voiceNames = keys(voices);
for i = 1:numel(voiceNames)
    name = voiceNames(i);
    params = voices(name);
    f0 = params(1); entry = params(2); exit = params(3);
    
    audio = audio + synthesizeVoice(t, f0, entry, exit, sr);
    fprintf("  %s: %.1f Hz (%.1fs → %.1fs)\n", name, f0, entry, exit);
end

audio = audio / max(abs(audio)) * 0.9;

% Save audio
audiowrite("demo_barbershop.wav", audio, sr);
fprintf("\n✓ Audio saved: demo_barbershop.wav\n");

%% Run tracker
fprintf("\n🔍 Running Multi-F0 Tracker...\n\n");

tracker = MultiF0Tracker( ...
    MaxVoices=4, ...
    MinFreq=80, ...
    MaxFreq=600, ...
    PeakThreshold=0.15, ...
    DetectExtraPitches=false);

result = tracker.track(audio, sr);

%% Filter short tracks (keep only substantive voices)
minFrames = 20;
result.SungVoices = result.SungVoices(cellfun(@(t) numel(t.Pitches) >= minFrames, result.SungVoices));

%% Visualize
fprintf("\n📊 Creating visualization...\n");

result.visualize( ...
    Title="Barbershop Quartet: A Dominant 7th (Just Intonation)", ...
    OutputPath="tracking_result.png");

%% Summary & export
result.summary();
result.exportCSV("tracking_data.csv");

fprintf("\n✅ Done!\n");

%% ===================== Helper Function =====================

function signal = synthesizeVoice(t, f0, entryTime, exitTime, sr)
    % Generate synthetic voice with vibrato and harmonics
    arguments
        t (:,1) double
        f0 (1,1) double
        entryTime (1,1) double
        exitTime (1,1) double
        sr (1,1) double
    end

    % Vibrato
    vibratoRate = 5;      % Hz
    vibratoDepth = 0.015; % 1.5%
    vibrato = vibratoDepth * sin(2*pi*vibratoRate*t);

    % Additive synthesis with formant-like weighting
    weights = [1.0, 0.7, 0.4, 0.3, 0.2, 0.15];
    signal = zeros(size(t));
    for h = 1:numel(weights)
        signal = signal + weights(h) * sin(2*pi*h*f0*t + h*vibrato);
    end

    % Amplitude envelope with smooth attack/release
    envelope = zeros(size(t));
    fadeLen = round(0.05 * sr);
    entryIdx = round(entryTime * sr) + 1;
    exitIdx = round(exitTime * sr);

    for i = 1:numel(t)
        if i < entryIdx
            envelope(i) = 0;
        elseif i < entryIdx + fadeLen
            envelope(i) = (i - entryIdx) / fadeLen;
        elseif i < exitIdx - fadeLen
            envelope(i) = 1;
        elseif i < exitIdx
            envelope(i) = (exitIdx - i) / fadeLen;
        end
    end

    envelope = smoothdata(envelope, "gaussian", 50);
    signal = signal .* envelope;
end

File: python/demo_barbershop.py
import numpy as np
import soundfile as sf
import os
import sys
from scipy.signal import savgol_filter

# Add current directory to path
sys.path.append(os.path.dirname(__file__))

from multi_f0_tracker import MultiF0Tracker

def synthesize_voice(t, f0, entry_time, exit_time, sr):
    """Generate synthetic voice with vibrato and harmonics."""
    # Vibrato
    vibrato_rate = 5.0      # Hz
    vibrato_depth = 0.015   # 1.5%
    vibrato = vibrato_depth * np.sin(2 * np.pi * vibrato_rate * t)

    # Additive synthesis with formant-like weighting
    weights = [1.0, 0.7, 0.4, 0.3, 0.2, 0.15]
    signal = np.zeros_like(t)
    for h_idx, weight in enumerate(weights):
        h = h_idx + 1
        signal += weight * np.sin(2 * np.pi * h * f0 * t + h * vibrato)

    # Amplitude envelope with smooth attack/release
    envelope = np.zeros_like(t)
    fade_len = int(0.05 * sr)
    entry_idx = int(entry_time * sr)
    exit_idx = int(exit_time * sr)

    # Apply fades
    if entry_idx + fade_len < len(t):
        envelope[entry_idx:entry_idx+fade_len] = np.linspace(0, 1, fade_len)
    
    if entry_idx + fade_len < exit_idx - fade_len:
        envelope[entry_idx+fade_len:exit_idx-fade_len] = 1.0
        
    if exit_idx - fade_len < len(t) and exit_idx - fade_len > 0:
        envelope[exit_idx-fade_len:exit_idx] = np.linspace(1, 0, fade_len)

    # Smooth the envelope
    # Using savgol_filter as a replacement for smoothdata(..., "gaussian")
    envelope = savgol_filter(envelope, 101, 3)
    envelope = np.clip(envelope, 0, 1)
    
    return signal * envelope

def main():
    sr = 22050
    duration = 9.5
    
    # Voice parameters: [frequency, entry_time, exit_time]
    voices = {
        "Bass":     [110.0,  0.5, 9.0],
        "Baritone": [137.5,  1.5, 8.0],
        "Lead":     [165.0,  2.5, 7.0],
        "Tenor":    [192.5,  3.5, 6.0]
    }

    print("🎵 Generating synthetic barbershop quartet (Python)...")
    
    t = np.linspace(0, duration, int(duration * sr), endpoint=False)
    audio = np.zeros_like(t)

    for name, params in voices.items():
        f0, entry, exit = params
        audio += synthesize_voice(t, f0, entry, exit, sr)
        print(f"  {name}: {f0:.1f} Hz ({entry:.1f}s → {exit:.1f}s)")

    # Normalize
    audio = audio / np.max(np.abs(audio)) * 0.9

    output_wav = "demo_barbershop_py.wav"
    sf.write(output_wav, audio, sr)
    print(f"\n✓ Audio saved: {output_wav}")

    print("\n🔍 Running Multi-F0 Tracker...")
    from multi_f0_tracker import TrackerConfig
    config = TrackerConfig(
        max_voices=4,
        min_freq=80,
        max_freq=600,
        peak_threshold=0.15,
        detect_extra_pitches=False
    )
    tracker = MultiF0Tracker(config=config)

    # track returns a dictionary
    raw_result = tracker.track(audio, sr)

    print("\n📊 Creating visualization...")
    from tracking_result import TrackingResult
    result = TrackingResult(
        times=raw_result["times"],
        voices=raw_result["sung_voices"],
        ghosts=raw_result["extra_pitches"],
        sr=raw_result["sample_rate"],
        hop=raw_result["hop_length"]
    )
    
    output_png = "tracking_result_py.png"
    result.visualize(title="Barbershop Quartet: A Dominant 7th (Just Intonation)", output_path=output_png)
    
    print(f"✓ Visualization saved: {output_png}")
    print("\n✅ Done!")

if __name__ == "__main__":
    # Ensure matplotlib works in headless environment
    import matplotlib
    matplotlib.use('Agg')
    main()
